### ISSUE TITLE
Nav Graph Scoped View Models?

### ISSUE THREAD

**Issue by nealsanche** [airbnb/MvRx#305]
MvRx has activity, fragment, parentFragment, and targetFragment scoped viewModel delegate functions. Not sure what the last one does, as I haven't seen an example.

Navigation Component has the ability to get a ViewModel that is scoped to a particular nested navigation graph. This is super handy when executing a flow that you need to do again and again, and don't want the viewModel to be reused between flows.

I know that using fragmentViewModel on the start fragment in the flow, and existingViewModel in the sub-parts of the flow can also do something similar to nav graph scoped view models.

But, where it gets fun is when you're trying to test fragments in isolation using something like FragmentScenario. existingViewModel will actually crash in this case because you might be testing the second fragment in a flow. Which actually brings to mind two specific questions.

How could a ViewModel be created while doing an Espresso test so that existingViewModel would succeed instead of crashing -- remember I'm testing fragments in isolation?

Is it possible to integrate nav graph scoped view models with MvRx in a simple way?

**Comment by elihart** [comment]
The testing system I'm working on with https://github.com/airbnb/MvRx/pull/282 would help to test fragments with existing view models, but I see how nav graph scope would still be helpful.

We don't use the navigation architecture component, so it's a little hard to comment on it. Based on https://developer.android.com/reference/kotlin/androidx/navigation/package-summary#navgraphviewmodels what you are doing seems reasonable.

we could maybe either expose an internal version of `mvrxFactory` to make this easier for people to use, or add a navigation module that provides this function.

It seems simple enough to use directly like this for now, but it is good to think about providing this from the library at some point. Thanks for sharing!

**Comment by marukami** [comment]
> It seems simple enough to use directly like this for now, but it is good to think about providing this from the library at some point. Thanks for sharing!

I spiked out a [prototype](https://github.com/airbnb/MvRx/compare/master...marukami:navigation-viewmodel-store) and it should be possible without changing any fundamentals to have something like this. 
```kotlin
private val viewModel: MyDetailViewModel by navGraphViewModel(R.id.my_graph_id)
``` 
 
@elihart would this approach make sense when navigation to the library?

**Comment by iMvBremen** [comment]
I think an easy fix to the problem would be to make [`ViewModelContext`](https://github.com/airbnb/MvRx/blob/9872c93d56a162cc09e9be37458bd046fa3f3446/mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelFactory.kt#L48) an `interface`/`abstract class` instead of a `sealed class`. This way we can scope ViewModels to whatever scope we like, e.g. a NavGraph or even a custom View/MvRxView, as long as it has its own `ViewModelStoreOwner` and `SavedStateRegistry`.

Only problem I see is that [`Bundle.toStateRestorer()`](https://github.com/airbnb/MvRx/blob/9872c93d56a162cc09e9be37458bd046fa3f3446/mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelProvider.kt#L100) casts given context to a specific `ViewModelContext` in order to call the `copy()` method. This is done to create a restored context, with restored arguments. For an interface, this could be rewritten to e.g. the following method on the `ViewModelContext` itself:
```kotlin
fun copyArgs(args: Any?): ViewModelContext
```

I'm not entirely sure if this is the way to go, so I'd love to hear some thoughts on this :)

**Comment by gpeal** [comment]
Fixed by https://github.com/airbnb/MvRx/pull/443 and will be released in the next 2.0 beta/rc
